### DISCUSSION TITLE
Process Component | failed when running in Laravel controller on Windows

### DISCUSSION THREAD

**Discussion by convers39** [symfony/symfony#42611] [Q&A]
I am working on a project using Laravel 6.20 on windows 10.

I need to run some commands from the controller, but somehow the command is always failing.

I am not sure why, here is what I have tried in the controller and the laravel tinker console:
```php
$process = new Process([...]); // either Process::fromShellCommandline($command);
$process->run();
```
I have tested those commands in the controller with Process component, tinker console, and windows native cmd. 
|command| controller | tinker console | native cmd|
|---|---|---|---|
|`cd path`| ✅ | ✅| ✅|
|`php -v`| ❌ | ✅ | ✅ |
|`schtasks`| ❌ | ✅ | ✅ |
|`ls`| ❌ | ✅ | ❌ |
|`.bat`file| ❌ | ✅ | ✅ |

I understand the command is by default passed to windows cmd.exe, however, I don't know why the `php -v` and `schtasks` are working well in tinker and native cmd, but failed on the Process component.

The error message from `getErrorOutput()` says those failed commands are 

**not recognized as an internal or external command, operable program, or batch file.** 

Does anyone have any idea on this issue? Or is that possible to use PowerShell instead of the default cmd to execute the commands? Really appreciate your help!!

**Comment by faizanakram99** [comment]
Make sure `PATH` is passed to your web server (nginx or apache) and also the user under which php process runs, has permission to execute those commands.

**Reply by convers39** [reply]
thank you for your reply.
I think I figure it out, it does related to the path.
Most commands work when I pass in the full path of that executable, 
e.g. `"C:\Program Files\xampp\php\php.exe" -v`
Some still have some issues, maybe relate with permission.

**Comment by convers39** [comment]
The answer is I need to use the absolute path. 

Got hint from [this answer](https://stackoverflow.com/questions/48321639/is-not-recognized-as-an-internal-or-external-command-operable-program-or-bat)

So this should work fine.
```
$process = new Process(['"C:\Program Files\xampp\php\php.exe"', '-v']); 
// either Process::fromShellCommandline('C:\Program Files\xampp\php\php.exe -v');
$process->run();
```

**Reply by faizanakram99** [reply]
Passing PATH env var to fast CGI or Apache automatically resolves executables to absolute path.

That's the idea.

You can't always use absolute path, application can be installed in any other location and you wouldn't want to update code every time
